### ISSUE TITLE
How to Upload Attachments via the BULK Api?

### ISSUE THREAD

**Issue by octym** [jsforce/jsforce#228]
While I can upload attachments just fine via CRUD, I have to upload about 1000 files - which would be insane to do via REST api.

I'm trying to do it with the BULK api, but I get this dread error: "**Binary field Body is only supported for content types ZIP_XML and ZIP_CSV**".

**Setup**
```javascript
var bulkInsert = function(sObject, records) {
	return new Promise(function(resolve, reject) {
		console.log('>>> Starting batch insert of', records.length, sObject, 'records ...');
		sfConn.bulk.load(sObject, 'insert', records, function(err, results) {
			if (err) {
				reject(err);
			} else {
				resolve(results);
			}
		});
	});
};
```

**Method #1:**
```javascript
bulkInsert('Attachment', [{
	ParentId: 'someSFId',
	Name: 'myImage',
	ContentType: 'image/jpeg',
	Body: Buffer(fs.readFileSync('/path/to/my/files'), 'binary').toString('base64')
}];
```

**Method #2:**
```javascript
bulkInsert('Attachment', [{
	ParentId: 'someSFId',
	Name: 'myImage',
	ContentType: 'image/jpeg',
	Body: '/path/to/my/jpg'
}];
```

They both fail with **Binary field Body is only supported for content types ZIP_XML and ZIP_CSV**.

However, the REST api works just fine ...
```javascript
sfConn.sobject('Attachment').create({
	ParentId: 'someSFId',
	Name: 'myImage',
	ContentType: 'image/jpeg',
	Body: Buffer(fs.readFileSync('/path/to/my/files'), 'binary').toString('base64')
});
```

How can I get the BULK api to work as well?

**Comment by stomita** [comment]
Attachment in bulk API is not supported yet.

**Comment by kylefarris** [comment]
I'm working on this feature in my fork (read: I actually have it working). It's really not that complicated to implement. I was just wondering how much of the functionality (zipping, etc...) should be incorporated into jsforce itself vs just showing how to do it in the documentation. The changes to jsforce itself were relatively minor (just allowing a user to choose the `contentType`), but if we wanted to have the full zipping capability built into the library, that would be a lot more extensive and would require a couple new dependencies (`adm-zip` and `streamifier`).

I'd like some thoughts from the jsforce team before making my pull request.

**Comment by stomita** [comment]
@kylefarris If you already have a implementation please give us the PR and we would like to check if it is suitable for merge to master.

**Comment by jkrinsky** [comment]
Hi! Any updates on this? @kylefarris could I get access to the fork you have this working on pretty please? :)

**Comment by kylefarris** [comment]
Dang... I don't think I ever pushed my commits to remote. I'll see if I still have them locally somewhere.

**Comment by kylefarris** [comment]
So, FWIW, I do still have the files so I've pushed them to my branch here on Github. I'm going through the process of merging the changes from the current Master as my branch is like 3 years old... Bit of a pain but not too bad. I should have it up shortly.

**Comment by lms15** [comment]
Yes will not open for me and not sure if package is downloaded or the theme?
Mamie

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by jkrinsky** [comment]
@kylefarris Thank you so much for graciously sharing your code with me! I have merged it into my local jsforce repo. Now I am trying to figure out how to download attachments via bulk.query and then upload them to a different Salesforce org using bulk.createJob() and job.createBatch(). Any assistance would be most appreciated!

**Comment by jkrinsky** [comment]
@kylefarris ok, figured out my stuff with the attachment downloading. Any way we could get you to share an example or two of creating attachments via the bulk api leveraging your changes? Thanks!

**Comment by lms15** [comment]
<?xml version="1.0" encoding="ISO-8859-1"?>
<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata"> <apiVersion>29.0</apiVersion> </ApexClass><?xml version="1.0" encoding="ISO-8859-1"?>
What does this mean?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________
